### PR TITLE
camera_cards work through 9/30

### PR DESCRIPTION
**Pull request by aeschweik**

### CODE REVIEW

File: camera_cards [hunk 1]
```diff
@@ -213,21 +216,24 @@ if [[ "${AIP}" == "YES" ]] ; then
...
     fi
+    STREAM_MD5="${METADATA_OUTPUT_DIR}/$(basename ${CONCATENATED_VIDEO_FILE})_bitstream_md5.txt"
     # create list of files to concatenate, formatted for ffmpeg
     while read FILE ; do
         echo "file '${FILE}'" >> "${TEMP_CONCATLIST}"
     done <"${TEMP_VIDEOLIST}"
     # concatenate video files in the order they are printed in $TEMP_CONCATLIST; map metadata from the first video file (in sequence) onto the concatenated file
-    ffmpeg -f concat -safe 0 -i "${TEMP_CONCATLIST}" -i "${FIRST_FILE}" -map 0 -map_metadata 1 -c copy "${CONCATENATED_VIDEO_FILE}"
-    _writelog "CONCATENATED VIDEO FILE" "${CONCATENATED_VIDEO_FILE}"
+    ffmpeg -f concat -safe 0 -i "${TEMP_CONCATLIST}" -i "${FIRST_FILE}" -map 0 -map_metadata 1 -c copy "${CONCATENATED_VIDEO_FILE}" -f md5 -map 0:a? -map 0:v -c copy "${STREAM_MD5}"
```
Comment (dericed) [inline camera_cards:225]: streamhash?

File: camera_cards [hunk 2]
```diff
@@ -56,18 +57,19 @@ _generate_metadata(){
...
         mediaconch -mi -ft "${FILE}" >> "${MEDIAINFO_OUTPUT}"
         exiftool "${FILE}" >> "${EXIFTOOL_OUTPUT}"
         ffprobe 2> /dev/null "${FILE}" -show_format -show_streams -show_data -show_error -show_versions -show_chapters -noprivate -of xml="q=1:x=1" > "${FFPROBE_OUTPUT}"
     done <"${METADATA_FILELIST}"
+    _writelog "Metadata reports generated" "$(date +%FT%T)"
 }
 
 # check for unassigned variables
-CAMERA_CARD_DIR="${1}"
+CAMERA_CARD_DIR="${1%/}"
```
Comment (dericed) [inline camera_cards:72]: why all the trailing slash finagling?

File: camera_cards [hunk 3]
```diff
@@ -312,13 +321,45 @@ if [[ "${TAR}" == "YES" ]] ; then
...
     _report -g "Creating tar archive..."
+    _writelog "Tar archive process process started" "$(date +%FT%T)"
     tar -czvf "${AIPDIR}/objects/${MEDIAID}.tar.gz" -C "${CAMERA_CARD_DIR}" .
     _writelog "TAR ARCHIVE" "${MEDIAID}.tar.gz"
+    _writelog "Tar archive process process finished" "$(date +%FT%T)"
+fi
+
+# error reporting
+_report -b "Checking for errors during the camera_cards process..."
+if [[ -z $(cat "${TREE}") ]] || [[ -z $(cat "${MEDIAINFO_OUTPUT}") ]] || [[ -z $(cat "${EXIFTOOL_OUTPUT}") ]] || [[ -z $(cat "${FFPROBE_OUTPUT}") ]] ; then
```
Comment (dericed) [inline camera_cards:332]: could use `[[ ! -s "$TREE" ]]` rather than cat'ing the whole file. Would be faster/cleaner.

File: camera_cards [hunk 4]
```diff
@@ -236,9 +242,11 @@ if [[ "${AIP}" == "YES" ]] ; then
         done <"${TEMP_AUDIOLIST}"
         ffmpeg -f concat -safe 0 -i "${TEMP_CONCATLIST}" -i "${FIRST_FILE}" -map 0 -map_metadata 1 -c copy "${CONCATENATED_AUDIO_FILE}"
         _writelog "CONCATENATED AUDIO FILE" "${CONCATENATED_AUDIO_FILE}"
+        _writelog "Audio concatenation process finished" "$(date +%FT%T)"
```
Comment (dericed) [inline camera_cards:245]: nbd, but maybe at some point, could at a `-t` option to `_writelog` to mean that the value of `key:value` should be a timestamp of now.

File: camera_cards [hunk 5]
```diff
@@ -312,13 +321,45 @@ if [[ "${TAR}" == "YES" ]] ; then
...
+fi
+if [[ "${FFMPEG_EXIT_CODE}" -gt 0 ]] || [[ "${FFMPEG_ERRORS}" -gt 0 ]] ; then
+    _report -r "FFmpeg may have encountered errors - review terminal output for more detail."
+else
+    _report -g "FFmpeg exit codes look ok"
+fi
+if [[ -s "${CONCATENATED_AUDIO_FILE}" ]] ; then
+    if [[ "${VIDEO_DURATION}" -ne "${AUDIO_DURATION}" ]] ; then
+    _report -r "Concatenated audio and video files were not the same length - review final merged file for sync at ${CONCATENATED_FILE} (video duration = ${VIDEO_DURATION}, audio duration = ${AUDIO_DURATION})"
+    _writelog "POSSIBLE_ERROR_REVIEW" "Concatenated audio and video files were not the same length"
```
Comment (dericed) [inline camera_cards:346]: What precision is the duration comparison at?  Like one NTSC frame with 48000 audio the closest duration match would be a video frame at 1001/30000 and audio at 1602/48000.
aka
0.03336666667 (video) versus 0.033375 (audio) is the closest duration alignment you can get. The next lower 480000 audio duration is at 1601/48000 or 0.03335416667 which is even further off.

File: camera_cards [hunk 6]
```diff
@@ -312,13 +321,45 @@ if [[ "${TAR}" == "YES" ]] ; then
...
+    else
+        _report -g "Concatenated audio and video process looks ok"
+    fi
+fi
+if [[ "${AIP}" == "YES" ]] ; then # check for losslessness of concatenated file by comparing bitstream md5s
+    OUTPUT_STREAM_MD5="$(_maketemp)"
+    ffmpeg -y -loglevel quiet -report -i "${CONCATENATED_VIDEO_FILE}" -f md5 -map 0:a? -map 0:v -c copy "${OUTPUT_STREAM_MD5}"
+    if [[ $(cat "${STREAM_MD5}") != $(cat "${OUTPUT_STREAM_MD5}") ]] ; then
+        _report -r "Concatenation process was not lossless! Review final merged file at ${CONCATENATED_VIDEO_FILE}"
```
Comment (dericed) [inline camera_cards:355]: streamhash could do this better, as this test would require the audio and video to be interleaved in the same order, which isn't required for lossless remuxing. Like `V1,A1,A2,V2,A3,A4` and `V1,A1,A2,A3,V2,A4` are lossless but in different interleavement patterns, streamhash comparisons would separate to one hash per stream.

Comment (aeschweik) [reply camera_cards:358]: Heya Dave, thanks! I agree that streamhash is better for this. Would you mind taking a look at this commit and letting me know if it makes sense: bcc5315f2a962870bb7d226c54eb611fddf56c45

File: camera_cards (hunk 1 above)
Comment (aeschweik) [reply camera_cards]: Changed in commit: bcc5315f2a962870bb7d226c54eb611fddf56c45

File: camera_cards (hunk 2 above)
Comment (aeschweik) [reply camera_cards]: Good point. changed in 53297f2b79e530ffcd57122ff29d2e6488306421

File: camera_cards (hunk 3 above)
Comment (aeschweik) [reply camera_cards]: 👍🏻 3e8dc3c8f0c3565fc409422589345afc5cdde55b

File: camera_cards (hunk 5 above)
Comment (aeschweik) [reply camera_cards:349]: Hmmm, good point. It's currently reporting the output of `mediainfo --Inform="Video;%Duration%"` and `mediainfo --Inform="Audio;%Duration%"`, so milliseconds. I was looking at dvpackager and it looks like it is using duration in seconds ([? hope I'm reading it right](https://github.com/mipops/dvrescue/blob/4e76bd384b8feec2e270ed8d42c87b5f2ba71dee/tools/dvpackager#L416)). Maybe better to round this to seconds?

File: camera_cards (hunk 4 above)
Comment (aeschweik) [reply camera_cards]: 👍🏻 859d8a3f2034021c316a069f7b2e763253cd4fce

**Review by dericed — APPROVED** [review]
Looks good to me. Can you recommend some steps for @BleakleyMcD to test the updates?

**Pull request merged**